compiler/rustc_passes/src/intrinsicck.rs

```diff
@@ -502,6 +506,37 @@ impl<'tcx> Visitor<'tcx> for ItemVisitor<'tcx> {
         ExprVisitor { tcx: self.tcx, param_env, typeck_results }.visit_body(body);
         self.visit_body(body);
     }
+
+    fn visit_inline_asm(&mut self, asm: &'tcx hir::InlineAsm<'tcx>, id: hir::HirId) {
+        for (op, op_sp) in asm.operands.iter() {
+            match *op {
+                // These are checked in ExprVisitor.
+                hir::InlineAsmOperand::In { .. }
+                | hir::InlineAsmOperand::Out { .. }
+                | hir::InlineAsmOperand::InOut { .. }
+                | hir::InlineAsmOperand::SplitInOut { .. } => {}
+                // No special checking is needed for these:
+                // - Typeck has checked that Const operands are integers.
+                // - AST lowering guarantees that SymStatic points to a static.
+                hir::InlineAsmOperand::Const { .. } | hir::InlineAsmOperand::SymStatic { .. } => {}
+                // Check that sym actually points to a function. Later passes
+                // depend on this.
+                hir::InlineAsmOperand::SymFn { anon_const } => {
+                    let ty = self.tcx.typeck_body(anon_const.body).node_type(anon_const.hir_id);
+                    match ty.kind() {
+                        ty::Never | ty::Error(_) => {}
+                        ty::FnDef(..) => {}
+                        _ => {
+                            self.tcx
+                                .sess
+                                .span_err(*op_sp, "asm `sym` operand must point to a fn or static");
```

Comment:
This would benefit from a structured diagnostic, first and foremost saying what kind of entity `anon_const` actually is. Something like (± other minor wording adjustments) is probably a more rustic diagnostic (cc @estebank if they have other opinions)

```
error: invalid `sym` operand
  --> $DIR/type-check-2.rs:88:19
   |
LL | global_asm!("{}", sym C);
   |                       ^ `C` is a constant
help: `sym` operands must refer to either a function or a static
```

Reply:
I agree for this and the other `struct_span_err` in `asm.rs`. It should be easy to do. Also, (nitpick) surround the keywords with backticks.